presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDiskPageInput.java

```diff
@@ -0,0 +1,147 @@
+import com.facebook.presto.spi.page.PagesSerde;
+import com.facebook.presto.spi.page.PagesSerdeUtil;
+import com.facebook.presto.spi.page.SerializedPage;
+import com.facebook.presto.spi.storage.TempDataOperationContext;
+import com.facebook.presto.spi.storage.TempStorage;
+import io.airlift.slice.InputStreamSliceInput;
+import org.apache.spark.SparkEnv;
+import org.apache.spark.storage.BlockManager;
+import org.apache.spark.storage.BroadcastBlockId;
+import scala.Option;
+import scala.reflect.ClassTag;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.spark.util.PrestoSparkUtils.computeChecksum;
+import static com.facebook.presto.spark.util.PrestoSparkUtils.toPrestoSparkSerializedPage;
+import static com.facebook.presto.spark.util.PrestoSparkUtils.toSerializedPage;
+import static java.util.Objects.requireNonNull;
+import static org.apache.spark.api.java.StorageLevels.MEMORY_AND_DISK;
+
+public class PrestoSparkDiskPageInput
+        implements PrestoSparkPageInput
+{
+    private static final Logger log = Logger.get(PrestoSparkDiskPageInput.class);
+
+    private final PagesSerde pagesSerde;
+    @GuardedBy("this")
+    private final List<Iterator<PrestoSparkSerializedPage>> serializedPageIterators;
+    @GuardedBy("this")
+    private int currentIteratorIndex;
+
+    public PrestoSparkDiskPageInput(PagesSerde pagesSerde,
+            List<List<PrestoSparkDiskPage>> broadcastTableFilesInfo,
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext)
+    {
+        this.pagesSerde = requireNonNull(pagesSerde, "pagesSerde is null");
+        this.serializedPageIterators = broadcastTableFilesInfo.stream()
+                .map(tableFiles -> loadBroadcastTableFromDisk(tableFiles, tempStorage, tempDataOperationContext))
+                .collect(Collectors.toList());
+    }
+
+    private synchronized Iterator<PrestoSparkSerializedPage> loadBroadcastTableFromDisk(
+            List<PrestoSparkDiskPage> broadcastTaskFilesInfo,
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext)
+    {
+        // Empty broadcast table
+        if (broadcastTaskFilesInfo.size() == 0) {
+            return Collections.emptyIterator();
+        }
+
+        List<PrestoSparkSerializedPage> pages = new ArrayList<>();
+        BlockManager blockManager = SparkEnv.get().blockManager();
+        BroadcastBlockId broadcastBlockId =
+                new BroadcastBlockId((broadcastTaskFilesInfo.get(0)).getBroadcastId(), "pos");
+
+        // Get the pages cached in blockmanger
+        Option<Object> object = blockManager.getSingle(broadcastBlockId, classTag(Object.class));
```

Comment:
Using block manager for caching is not ideal, as Presto on Spark doesn't cooperate with the Spark memory manager. Thus upon memory pressure the pages will never get spilled to disk.

Ideally we should cache the hash table itself internally (in the HashBuilderOperator). But that will be our next iteration.

For now I would recommend you caching the broadcasted pages into a local temporary file directly, bypassing the BlockManager. The system level file system cache will do the memory optimization automatically (if there's free memory available).

Reply:
As per offline discussions, we will remove caching of pages in blockmanager. Instead, we will use an in-memory cache within PoS which will cache HT consumed by the recent stage.